fcrepo/fcrepo
# Lock ghost nodes when creating a resource

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3584

# What does this Pull Request do?
Locks ghost nodes above a resource that is being created so another transaction cannot replace them while the transaction is not committed.

# How should this be tested?

Before PR:
1. Create a transaction.
     ```
      > curl -i -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/fcr:tx -X POST

     HTTP/1.1 201 Created
     ...
     Location: http://localhost:8080/rest/fcr:tx/a5912bd9-a479-4e23-9c97-bc4e85f59539
     Link: <http://localhost:8080/rest/fcr:tx/a5912bd9-a479-4e23-9c97-bc4e85f59539>; rel="http://fedora.info/definitions/v4/transaction#commitEndpoint"
     ```
1. Create a binary with ghost nodes inside the transaction.
     ```
     > curl -i -u fedoraAdmin:fedoraAdmin -H "Atomic-ID: http://localhost:8080/rest/fcr:tx/a5912bd9-a479-4e23-9c97-bc4e85f59539" -X PUT -H "Content-type: text/plain" --data-binary "SOME CONTENT" http://localhost:8080/rest/a/b/c

     HTTP/1.1 201 Created
     ...

     http://localhost:8080/rest/a/b/c
     ```
1. Outside the transaction create a binary at one of the ghost nodes.
      ```
      > curl -i -u fedoraAdmin:fedoraAdmin -X PUT -H "Content-type: text/plain" --data-binary "SOME OTHER CONTENT" http://localhost:8080/rest/a

     HTTP/1.1 201 Created
     ...

     http://localhost:8080/rest/a
     ```
1. Commit the transaction 
     ```
     > curl -i -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/fcr:tx/a5912bd9-a479-4e23-9c97-bc4e85f59539 -XPUT
      
     HTTP/1.1 204 No Content
     ```
1. See that a binary has a binary child
     ```
     > curl -u fedoraAdmin:fedoraAdmin -i http://localhost:8080/rest/a

     HTTP/1.1 200 OK
     ...
     Content-Type: text/plain
     Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel="type"

     SOME OTHER CONTENT
     ```
     ```
     > curl -u fedoraAdmin:fedoraAdmin -i http://localhost:8080/rest/a/b/c

     HTTP/1.1 200 OK
     ...
     Content-Type: text/plain
     Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel="type"

     SOME CONTENT
     ```

Pull in this PR, re-do the above steps. Note the divergence at step 3

3. Outside the transaction create a binary at one of the ghost nodes.
      ```
      > curl -i -u fedoraAdmin:fedoraAdmin -X PUT -H "Content-type: text/plain" --data-binary "SOME OTHER CONTENT" http://localhost:8080/rest/a

     HTTP/1.1 409 Conflict
     Date: Thu, 25 Nov 2021 14:57:04 GMT
     Set-Cookie: JSESSIONID=node0ms1ix09jq0ag1m2863u0x1pn02.node0; Path=/
     Expires: Thu, 01 Jan 1970 00:00:00 GMT
     Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Wed, 24-Nov-2021 14:57:04 GMT; SameSite=lax
     Content-Type: text/plain;charset=utf-8
     Content-Length: 79
     Server: Jetty(9.4.44.v20210927)

     Cannot update info:fedora/a because it is being updated by another transaction.
     ```
# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers
